### PR TITLE
WIP: Allocator- and fallibility-polymorphic collections

### PR DESCRIPTION
**Pull request by Ericson2314**
This picks up where https://github.com/rust-lang/rust/pull/50882 will leave off, adding allocator parameters to the other collections. It also adds and associated error type for allocation, so as to enable fallibility-polymorphic code and support the few `impl`s and other functions which require "infallible" allocation.

When done, should convert all the collections for https://github.com/rust-lang/rust/issues/42774

---

Unfortunately, as documented in https://github.com/rust-lang/rust/issues/52396, I had to make an `AllocHelper` trait in order to get my `default impl` to work.

~~Also, `OK(_): Result<T, !>` is no longer deemed and infallible pattern so my `let Ok(..) = ..;` doesn't work. I can switch it to something else, but I'd like to make sure that's intentional first.~~ *edit* fixed with `exhaustive_patterns`.

### CODE REVIEW

File: src/libcore/alloc.rs
Comment (bjorn3) [inline src/libcore/alloc.rs]: The last word of this sentence is missing.

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: src/libcore/alloc.rs
Comment (Ericson2314) [reply src/libcore/alloc.rs]: Thanks :)

**Comment by alexcrichton** [comment]
r? @alexcrichton 

Thanks for the PR @Ericson2314! I think at this time though the libs team doesn't have the bandwidth to review and land this before the 2018 edition. This is a very large step for us to take which we want to be sure to take carefully. Would it be possible to postpone this until after the 2018 edition has released?

**Comment by Ericson2314** [comment]
I suppose that's fine. When is that likely to be?

I'd just like the fact that there's work done towards this to affect decisions like Alloc stabilization. (I don't think I actually need to break any APIs, but let this show that many long-standing issues with alloc are in fact resolvable, so who knows what else a deep dive would turn up.)

**Comment by Ericson2314** [comment]
If at least https://github.com/rust-lang/rust/pull/50882 can not be blocked, that would help me keep this up-to-date in the meantime.

**Comment by alexcrichton** [comment]
https://github.com/rust-lang/rust/pull/50882 AFAIK is largely technically blocked on the regression in error messages, but it's basically in the same boat as this PR in that it requires some discussion and design which the libs team doesn't have the bandwidth for right now, but after the edition 2018 release we can revisit.

**Comment by glandium** [comment]
#50882 is not blocked on regression in error messages anymore. It's blocked on one test failing in what looks like a race condition, and that only happens when building with LLVM 5.

**Comment by alexcrichton** [comment]
@glandium er sorry but that PR is also large enough and contentious enough that we're not going to be able to land it before the edition, I'll comment over there though.

**Comment by SimonSapin** [comment]
Making `AllocErr` generic a type parameter / associated type seems like a significant enough design change in public APIs to merit being discussed outside of a large PR.

**Comment by Aaronepower** [comment]
Triage: Based on the comments above I'm closing this PR and marking it `S-blocked-closed` on the landing of the 2018 edition. Feel free to reopen once the 2018 edition lands!

**Comment by lzutao** [comment]
Now the 2018 edition is landed. The PR could be reopened.

**Comment by Ericson2314** [comment]
@lzutao thanks for you interest, but this depends on https://github.com/rust-lang/rust/pull/50882, which is still blocked on a bug in an older LLVM that we still try to support. Please go read that thread for details.

**Pull request closed without merging**
